### PR TITLE
Roop-multi changed the implementation of multi-threading processing for nvidia GPU.

### PR DESCRIPTION
**Pull request by Moeblack**
I tried next branch, but I thought the multi-thread processing is not going well with my Nvidia GPU.
https://github.com/microsoft/onnxruntime/discussions/8984
in this issue, maintainer said that the Parallel executor is not reliable for Nvidia GPU, so I modified the Code.

If you thought it's useful, I thought It's also can be adapted for AMD GPU. after modifying Code.
after the changing, I got 100% performence improvement on Nvidia 2080Super.
IF you have a better GPU, you can change the Thread number of this program. it will be faster I guess.

original Next branch
image
roop-multi
image
 
on the same demo video.

### CODE REVIEW

File: roop/swapper.py [hunk 1]
```diff
@@ -49,18 +64,33 @@ def process_video(source_img, frame_paths, preview_callback):
...
-                result = process_faces(source_face, frame, progress)
-                cv2.imwrite(frame_path, result)
-                if preview_callback:
-                    preview_callback(cv2.cvtColor(result, cv2.COLOR_BGR2RGB))                
-            except Exception:
-                pass
-            progress.update(1)
+        # nvidia multi-threading
+        if roop.globals.gpu_vendor == 'nvidia':
+            progress.set_postfix(cuda_utilization="{:02d}%".format(torch.cuda.utilization()), cuda_memory="{:02d}GB".format(torch.cuda.memory_usage()))
```
Comment (henryruhs) [inline roop/swapper.py:69]: Move to multi_process_frames() method

File: roop/swapper.py [hunk 2]
```diff
@@ -49,18 +64,33 @@ def process_video(source_img, frame_paths, preview_callback):
...
-            try:
-                result = process_faces(source_face, frame, progress)
-                cv2.imwrite(frame_path, result)
-                if preview_callback:
-                    preview_callback(cv2.cvtColor(result, cv2.COLOR_BGR2RGB))                
-            except Exception:
-                pass
-            progress.update(1)
+        # nvidia multi-threading
+        if roop.globals.gpu_vendor == 'nvidia':
```
Comment (henryruhs) [inline roop/swapper.py:68]: I suggest you go for if `roop.globals.gpu_vendor is not None`

File: roop/swapper.py [hunk 3]
```diff
@@ -4,24 +4,30 @@
...
-        else:
-            session_options.intra_op_num_threads = roop.globals.cpu_threads
-        session_options.execution_mode = onnxruntime.ExecutionMode.ORT_PARALLEL
-        model_path = os.path.join(os.path.abspath(os.path.dirname(__file__)), '../inswapper_128.onnx')
-        FACE_SWAPPER = insightface.model_zoo.get_model(model_path, providers=roop.globals.providers, session_options=session_options)
+    with lock:
+        if FACE_SWAPPER is None:
+            model_path = os.path.join(os.path.abspath(os.path.dirname(__file__)), '../inswapper_128.onnx')
+            # If use nvidia-gpu, change implementation method.
+            if roop.globals.gpu_vendor == 'nvidia':
```
Comment (henryruhs) [inline roop/swapper.py:21]: I suggest you go for if `roop.globals.gpu_vendor is not None`

File: roop/swapper.py (hunk 1 above)
Comment (henryruhs) [inline roop/swapper.py:69]: Remove the torch imports and  torch usage as AMD does not like it

**Review by henryruhs — CHANGES_REQUESTED** [review]

**Review by henryruhs — APPROVED** [review]

**Comment by cuong29102003** [comment]
i got this error could you tell me how to fix it, can i have U discord ... for help me 
image

**Pull request merged**